File: sdk/python/src/spinneret/watcher.py
"""Thread-based config watcher with long polling and local snapshots."""

from __future__ import annotations

import logging
import threading
import time
from collections.abc import Callable, Iterable
from os import PathLike
from pathlib import Path
from types import TracebackType
from typing import TYPE_CHECKING, Optional, Union

from . import _calls as calls
from ._calls import ConfigKeyLike
from ._retry import NO_RETRY, is_retryable_background_error
from ._snapshot import SecretPredicate, SnapshotStore
from ._watch_core import WatchOptions, WatchState
from .errors import REASON_CLIENT_CLOSED, SpinneretError
from .models import ConfigItem, WatchItem

if TYPE_CHECKING:
    from .client import Client

__all__ = ["ChangeCallback", "ConfigWatcher"]

logger = logging.getLogger("spinneret.config")

#: Callback invoked with every changed config item.
ChangeCallback = Callable[[ConfigItem], None]

_STOP_TIMEOUT = 1.0


class ConfigWatcher:
    """Keeps a set of config items up to date with ``ConfigService/WatchConfig``.

    :meth:`start` loads the items with ``BatchGetConfig``; when the server is
    unavailable it falls back to the local snapshots. A daemon thread then
    long-polls for changes, stores the latest items, writes snapshots
    atomically and invokes the change callbacks. Callbacks run on the thread
    that applied the change: the caller of :meth:`start` for the initial
    values, the watcher thread afterwards. They must not block for long.

    Secret items are kept in memory only, unless ``cache_secrets=True``
    (requires ``cryptography``), in which case their snapshots are encrypted
    with a key derived from the API token. An item is secret when the server
    sets ``has_secret_refs`` (its published content contained resolved
    ``${secret:...}`` references) or when ``treat_as_secret`` returns true.
    Use the predicate for items whose content embeds sensitive values
    directly instead of referencing secrets.

    Example::

        with client.config_watcher(["crawler/search.json"]) as watcher:
            item = watcher.get("crawler", "search.json")
    """

    def __init__(
        self,
        client: Client,
        items: Iterable[ConfigKeyLike],
        *,
        namespace: str = "",
        timeout_ms: int = 30_000,
        cache_dir: Optional[Union[str, PathLike[str]]] = None,
        snapshots: bool = True,
        cache_secrets: bool = False,
        treat_as_secret: Optional[SecretPredicate] = None,
        on_change: Optional[ChangeCallback] = None,
        options: Optional[WatchOptions] = None,
    ) -> None:
        """Create a watcher (not started).

        Args:
            client: Client used for the calls.
            items: Items to watch: :class:`ConfigKey`, ``(group, key)`` or ``"group/key"``.
            namespace: Namespace name; empty means the token namespace.
            timeout_ms: Long-poll timeout (1..60000); ignored when ``options`` is given.
            cache_dir: Snapshot directory (default: the client's ``cache_dir``).
            snapshots: Whether to read and write local snapshots.
            cache_secrets: Encrypt and persist secret items instead of skipping them.
            treat_as_secret: Predicate marking items as secret in addition to
                those flagged with ``has_secret_refs``, e.g.
                ``lambda item: item.group == "signing"``.
            on_change: Callback registered with :meth:`add_listener`.
            options: Advanced loop options.

        Raises:
            ConfigurationError: ``cache_secrets`` is set but ``cryptography`` is missing.
            ValueError: No items or more than 200 items were given.
        """
        self._client = client
        self._namespace = namespace
        self._options = options or WatchOptions(timeout_ms=timeout_ms)
        self._state = WatchState(calls.to_config_keys(items))
        settings = client.settings
        self._store: Optional[SnapshotStore] = None
        if snapshots:
            self._store = SnapshotStore(
                Path(cache_dir) if cache_dir is not None else settings.cache_dir,
                host=settings.host,
                namespace=namespace,
                token=settings.token,
                cache_secrets=cache_secrets,
                treat_as_secret=treat_as_secret,
            )
        self._listeners: list[ChangeCallback] = [on_change] if on_change is not None else []
        self._cond = threading.Condition()
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._loaded = False
        self._from_snapshot = False

    @property
    def from_snapshot(self) -> bool:
        """Whether the current items come from local snapshots (server not confirmed yet)."""
        with self._cond:
            return self._from_snapshot

    @property
    def running(self) -> bool:
        """Whether the watch thread is running."""
        return self._thread is not None and self._thread.is_alive()

    @property
    def snapshot_store(self) -> Optional[SnapshotStore]:
        """Snapshot store, or ``None`` when snapshots are disabled."""
        return self._store

    def get(self, group: str, key: str) -> Optional[ConfigItem]:
        """Latest known item, or ``None`` when it is unknown or missing."""
        with self._cond:
            return self._state.get(group, key)

    def items(self) -> dict[tuple[str, str], ConfigItem]:
        """Copy of all known items keyed by ``(group, key)``."""
        with self._cond:
            return self._state.items()

    def add_listener(self, callback: ChangeCallback) -> None:
        """Register a callback invoked with every changed item."""
        with self._cond:
            self._listeners = [*self._listeners, callback]

    def remove_listener(self, callback: ChangeCallback) -> None:
        """Unregister a callback (no-op when unknown)."""
        with self._cond:
            self._listeners = [cb for cb in self._listeners if cb is not callback]

    def start(self, *, wait: bool = True) -> ConfigWatcher:
        """Load the items and start the watch thread.

        Args:
            wait: Perform the initial load before returning (errors other than
                server unavailability are raised). When ``False`` the load
                happens in the background thread.

        Raises:
            SpinneretError: The initial load failed for a reason other than
                server unavailability (for example ``permission_denied``).
        """
        if self._stop.is_set():
            raise RuntimeError("a stopped ConfigWatcher cannot be restarted")
        if self._thread is not None:
            return self
        if wait:
            self._initial_load()
        thread = threading.Thread(target=self._run, name="spinneret-config-watcher", daemon=True)
        self._thread = thread
        thread.start()
        return self

    def stop(self, timeout: float = _STOP_TIMEOUT) -> None:
        """Stop the watch loop.

        An in-flight long poll cannot be interrupted; the thread exits when it
        returns (callbacks are not invoked after ``stop``).
        """
        self._stop.set()
        with self._cond:
            self._cond.notify_all()
        thread = self._thread
        if thread is not None and thread is not threading.current_thread():
            thread.join(timeout)

    def wait_for_change(
        self,
        group: Optional[str] = None,
        key: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> Optional[ConfigItem]:
        """Block until an item (optionally a specific one) changes after this call.

        Returns:
            The changed item, or ``None`` on timeout or when the watcher stops.
        """
        deadline = None if timeout is None else time.monotonic() + timeout
        with self._cond:
            start = self._state.sequence
            while True:
                item = self._state.changed_since(start, group, key)
                if item is not None:
                    return item
                if self._stop.is_set():
                    return None
                remaining = None if deadline is None else deadline - time.monotonic()
                if remaining is not None and remaining <= 0:
                    return None
                self._cond.wait(remaining)

    def __enter__(self) -> ConfigWatcher:
        return self.start()

    def __exit__(
        self,
        exc_type: Optional[type[BaseException]],
        exc: Optional[BaseException],
        tb: Optional[TracebackType],
    ) -> None:
        self.stop()

    # -- internals ----------------------------------------------------------

    def _initial_load(self) -> None:
        keys = self._state.keys
        try:
            response = self._client.batch_get_config(keys, namespace=self._namespace)
        except SpinneretError as exc:
            if not is_retryable_background_error(exc):
                raise
            logger.warning("spinneret config server unavailable, using local snapshots: %s", exc)
            loaded = []
            if self._store is not None:
                loaded = [
                    item
                    for item in (self._store.load(k.group, k.key) for k in keys)
                    if item is not None
                ]
            self._apply(loaded, persist=False, from_snapshot=True)
        else:
            if response.missing:
                logger.info(
                    "spinneret config items not found: %s",
                    ", ".join(f"{k.group}/{k.key}" for k in response.missing),
                )
            self._apply(response.items, persist=True, from_snapshot=False)
        self._loaded = True

    def _apply(self, items: Iterable[ConfigItem], *, persist: bool, from_snapshot: bool) -> None:
        with self._cond:
            changed = self._state.apply(items)
            self._from_snapshot = from_snapshot
            listeners = list(self._listeners)
            self._cond.notify_all()
        if persist and self._store is not None:
            for item in changed:
                self._store.save(item)
        if self._stop.is_set():
            return
        for item in changed:
            for callback in listeners:
                try:
                    callback(item)
                except Exception:
                    logger.exception(
                        "spinneret config change callback raised for %s/%s", item.group, item.key
                    )

    def _run(self) -> None:
        failures = 0
        while not self._stop.is_set():
            try:
                if not self._loaded:
                    self._initial_load()
                    continue
                call = calls.watch_config(
                    self._namespace,
                    self._state_watch_items(),
                    self._options.timeout_ms,
                    self._client.timeouts.watch_grace,
                )
                response = self._client._call(call, retry=NO_RETRY)
            except Exception as exc:
                if self._stop.is_set():
                    return
                if isinstance(exc, SpinneretError) and exc.reason == REASON_CLIENT_CLOSED:
                    logger.info("spinneret config watcher stopped: the client has been closed")
                    self._stop.set()
                    with self._cond:
                        self._cond.notify_all()
                    return
                delay = self._options.backoff.delay(failures)
                failures += 1
                if isinstance(exc, SpinneretError) and is_retryable_background_error(exc):
                    logger.warning(
                        "spinneret config watch failed, retrying in %.1fs: %s", delay, exc
                    )
                else:
                    logger.error(
                        "spinneret config watch failed, retrying in %.1fs", delay, exc_info=exc
                    )
                self._stop.wait(delay)
                continue
            failures = 0
            self._apply(response.items, persist=True, from_snapshot=False)

    def _state_watch_items(self) -> list[WatchItem]:
        with self._cond:
            return self._state.watch_items()

File: sdk/go/spinneret/watcher.go
package spinneret

import (
	"context"
	"log/slog"
	"maps"
	"runtime/debug"
	"slices"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"connectrpc.com/connect"
)

// Default config watcher tuning.
const (
	DefaultWatchTimeout      = 30 * time.Second
	DefaultWatchBackoff      = time.Second
	DefaultWatchMaxBackoff   = 30 * time.Second
	watchMinPollInterval     = 200 * time.Millisecond
	maxConfigGroupLength     = 128
	maxConfigKeyLength       = 256
	maxConfigNamespaceLength = 64
)

// ConfigKey addresses a config item inside the namespace.
type ConfigKey struct {
	Group string
	Key   string
}

// ParseConfigKey parses "group/key" (the group ends at the first slash).
func ParseConfigKey(s string) (ConfigKey, error) {
	group, key, ok := strings.Cut(s, "/")
	k := ConfigKey{Group: group, Key: key}
	if !ok {
		return k, newError(connect.CodeInvalidArgument, ReasonInvalidArgument, "config key %q must look like group/key", s)
	}
	return k, k.validate()
}

// String returns "group/key".
func (k ConfigKey) String() string { return k.Group + "/" + k.Key }

func (k ConfigKey) validate() error {
	if k.Group == "" || len([]rune(k.Group)) > maxConfigGroupLength {
		return newError(connect.CodeInvalidArgument, ReasonInvalidArgument, "config group must be 1..%d characters", maxConfigGroupLength)
	}
	if k.Key == "" || len([]rune(k.Key)) > maxConfigKeyLength {
		return newError(connect.CodeInvalidArgument, ReasonInvalidArgument, "config key must be 1..%d characters", maxConfigKeyLength)
	}
	return nil
}

// WatcherOptions configures a [ConfigWatcher].
type WatcherOptions struct {
	// Items are the watched items (1..200, duplicates are ignored).
	Items []ConfigKey
	// Namespace is the namespace name; empty means the token namespace.
	Namespace string
	// Timeout is the long-poll wait sent to the server (default 30s, max 60s).
	Timeout time.Duration
	// SnapshotDir enables local snapshots under this directory: fetched items
	// are written atomically and loaded when the server is unavailable at
	// start. Items with has_secret_refs, content with an unresolved
	// "${secret:" reference and items matched by TreatAsSecret are never
	// written (older snapshots of them are removed). Empty disables snapshots.
	SnapshotDir string
	// TreatAsSecret marks additional items as secret, e.g. items embedding
	// sensitive values directly. It cannot exempt items flagged by the server;
	// a predicate that panics counts as true.
	TreatAsSecret func(*ConfigItem) bool
	// OnChange is registered as the first change listener.
	OnChange func(*ConfigItem)
	// InitialBackoff is the first delay after a failed poll (default 1s).
	InitialBackoff time.Duration
	// MaxBackoff caps the delay between failed polls (default 30s).
	MaxBackoff time.Duration
}

// ConfigWatcher keeps a set of config items up to date with
// ConfigService/WatchConfig.
//
// Start loads the items with BatchGetConfig (falling back to local snapshots
// when the server is unavailable) and starts a goroutine that long-polls for
// changes until the Start context ends, Stop is called or the client is
// closed. Every change is stored, written to the snapshot directory and passed
// to the listeners: on the goroutine calling Start for the initial values, on
// the watch goroutine afterwards. Listeners run sequentially, must not block
// for long and must treat items as read-only; a panicking listener is logged
// and does not stop the watcher.
type ConfigWatcher struct {
	client  *Client
	opts    WatcherOptions
	keys    []ConfigKey
	watched map[ConfigKey]struct{}
	store   *snapshotStore
	logger  *slog.Logger
	rnd     func() float64

	dispatching atomic.Bool

	mu           sync.Mutex
	items        map[ConfigKey]*ConfigItem
	changedSeq   map[ConfigKey]uint64
	seq          uint64
	listeners    []func(*ConfigItem)
	fromSnapshot bool
	started      bool
	stopped      bool
	changed      chan struct{}
	cancel       context.CancelFunc
	done         chan struct{}
}

// NewConfigWatcher creates a watcher bound to the client (not started).
func (c *Client) NewConfigWatcher(opts WatcherOptions) (*ConfigWatcher, error) {
	if c.closed.Load() {
		return nil, newError(connect.CodeFailedPrecondition, ReasonClientClosed, "the client has been closed")
	}
	w := &ConfigWatcher{
		client:     c,
		watched:    make(map[ConfigKey]struct{}),
		logger:     c.s.logger,
		rnd:        c.rnd,
		items:      make(map[ConfigKey]*ConfigItem),
		changedSeq: make(map[ConfigKey]uint64),
		changed:    make(chan struct{}),
		done:       make(chan struct{}),
	}
	if err := w.configure(opts); err != nil {
		return nil, err
	}
	if opts.SnapshotDir != "" {
		w.store = newSnapshotStore(opts.SnapshotDir, c.s.host, opts.Namespace, opts.TreatAsSecret, c.s.logger)
	}
	if opts.OnChange != nil {
		w.listeners = []func(*ConfigItem){opts.OnChange}
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return nil, newError(connect.CodeFailedPrecondition, ReasonClientClosed, "the client has been closed")
	}
	c.watchers[w] = struct{}{}
	return w, nil
}

func (w *ConfigWatcher) configure(opts WatcherOptions) error {
	for _, k := range opts.Items {
		if err := k.validate(); err != nil {
			return err
		}
		if _, dup := w.watched[k]; !dup {
			w.watched[k] = struct{}{}
			w.keys = append(w.keys, k)
		}
	}
	if len(w.keys) == 0 || len(w.keys) > MaxConfigItems {
		return errInvalidOptions("a config watcher needs 1..%d items", MaxConfigItems)
	}
	if len([]rune(opts.Namespace)) > maxConfigNamespaceLength {
		return errInvalidOptions("namespace must be at most %d characters", maxConfigNamespaceLength)
	}
	if opts.Timeout == 0 {
		opts.Timeout = DefaultWatchTimeout
	}
	if opts.Timeout < time.Millisecond || opts.Timeout > MaxWatchTimeoutMs*time.Millisecond {
		return errInvalidOptions("watch timeout must be within 1ms..60s")
	}
	if opts.InitialBackoff < 0 || opts.MaxBackoff < 0 {
		return errInvalidOptions("watch backoff must not be negative")
	}
	if opts.InitialBackoff == 0 {
		opts.InitialBackoff = DefaultWatchBackoff
	}
	if opts.MaxBackoff == 0 {
		opts.MaxBackoff = DefaultWatchMaxBackoff
	}
	opts.MaxBackoff = max(opts.MaxBackoff, opts.InitialBackoff)
	w.opts = opts
	return nil
}

// Keys returns the watched keys in declaration order.
func (w *ConfigWatcher) Keys() []ConfigKey { return slices.Clone(w.keys) }

// Get returns the latest known item.
func (w *ConfigWatcher) Get(group, key string) (*ConfigItem, bool) {
	w.mu.Lock()
	defer w.mu.Unlock()
	item, ok := w.items[ConfigKey{Group: group, Key: key}]
	return item, ok
}

// Items returns a copy of the map of known items.
func (w *ConfigWatcher) Items() map[ConfigKey]*ConfigItem {
	w.mu.Lock()
	defer w.mu.Unlock()
	return maps.Clone(w.items)
}

// Versions returns the known version of every item (items never published are absent).
func (w *ConfigWatcher) Versions() map[ConfigKey]int32 {
	w.mu.Lock()
	defer w.mu.Unlock()
	out := make(map[ConfigKey]int32, len(w.items))
	for k, item := range w.items {
		out[k] = item.GetVersion()
	}
	return out
}

// FromSnapshot reports whether the items come from local snapshots and the
// server has not answered yet.
func (w *ConfigWatcher) FromSnapshot() bool {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.fromSnapshot
}

// AddListener registers a callback invoked with every changed item.
func (w *ConfigWatcher) AddListener(fn func(*ConfigItem)) {
	if fn == nil {
		return
	}
	w.mu.Lock()
	defer w.mu.Unlock()
	w.listeners = append(slices.Clip(w.listeners), fn)
}

// Done returns a channel closed when the watch goroutine has exited.
func (w *ConfigWatcher) Done() <-chan struct{} { return w.done }

// Start loads the items and starts the watch goroutine, which runs until ctx
// ends, Stop is called or the client is closed.
//
// When the server is unavailable ([IsRetryable] errors) the snapshots are
// loaded instead and FromSnapshot reports true until the server answers.
// Other errors (for example permission_denied) are returned and the watcher
// can be started again.
func (w *ConfigWatcher) Start(ctx context.Context) error {
	w.mu.Lock()
	switch {
	case w.stopped:
		w.mu.Unlock()
		return newError(connect.CodeFailedPrecondition, ReasonFailedPrecondition, "a stopped config watcher cannot be restarted")
	case w.started:
		w.mu.Unlock()
		return newError(connect.CodeFailedPrecondition, ReasonFailedPrecondition, "config watcher already started")
	}
	w.started = true
	w.mu.Unlock()
	if err := w.initialLoad(ctx); err != nil {
		w.mu.Lock()
		w.started = false
		if w.stopped {
			// Stop ran during the load and left closing Done to Start.
			close(w.done)
		}
		w.mu.Unlock()
		return err
	}
	loopCtx, cancel := context.WithCancel(ctx)
	w.mu.Lock()
	if w.stopped {
		w.mu.Unlock()
		cancel()
		close(w.done)
		return newError(connect.CodeFailedPrecondition, ReasonFailedPrecondition, "config watcher stopped during start")
	}
	w.cancel = cancel
	w.mu.Unlock()
	go w.run(loopCtx)
	return nil
}

// Stop ends the watch loop, cancelling an in-flight long poll, and waits for
// the goroutine to exit (unless called from a listener). Stop is idempotent.
func (w *ConfigWatcher) Stop() {
	w.mu.Lock()
	alreadyStopped := w.stopped
	w.stopped = true
	cancel := w.cancel
	started := w.started
	w.broadcastLocked()
	w.mu.Unlock()
	if !alreadyStopped {
		w.client.forgetWatcher(w)
	}
	if cancel == nil {
		// Not running: close Done unless a Start in progress will do it.
		if !alreadyStopped && !started {
			close(w.done)
		}
		return
	}
	cancel()
	if !w.dispatching.Load() {
		<-w.done
	}
}

// WaitForChange blocks until an item changes after the call and returns it.
// Empty group or key match any value. It fails when ctx ends or the watcher
// stops.
func (w *ConfigWatcher) WaitForChange(ctx context.Context, group, key string) (*ConfigItem, error) {
	w.mu.Lock()
	start := w.seq
	for {
		if item := w.changedSinceLocked(start, group, key); item != nil {
			w.mu.Unlock()
			return item, nil
		}
		if w.stopped {
			w.mu.Unlock()
			return nil, newError(connect.CodeFailedPrecondition, ReasonFailedPrecondition, "config watcher stopped")
		}
		changed := w.changed
		w.mu.Unlock()
		select {
		case <-changed:
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		w.mu.Lock()
	}
}

func (w *ConfigWatcher) changedSinceLocked(start uint64, group, key string) *ConfigItem {
	var (
		best    ConfigKey
		bestSeq uint64
	)
	for k, seq := range w.changedSeq {
		if seq <= start || (group != "" && k.Group != group) || (key != "" && k.Key != key) {
			continue
		}
		if bestSeq == 0 || seq < bestSeq {
			best, bestSeq = k, seq
		}
	}
	if bestSeq == 0 {
		return nil
	}
	return w.items[best]
}

func (w *ConfigWatcher) broadcastLocked() {
	close(w.changed)
	w.changed = make(chan struct{})
}

func (w *ConfigWatcher) refs() []*ConfigRef {
	refs := make([]*ConfigRef, len(w.keys))
	for i, k := range w.keys {
		refs[i] = &ConfigRef{Group: k.Group, Key: k.Key}
	}
	return refs
}

func (w *ConfigWatcher) initialLoad(ctx context.Context) error {
	resp, err := w.client.BatchGetConfig(ctx, &BatchGetConfigRequest{Namespace: w.opts.Namespace, Items: w.refs()})
	if err != nil {
		if !IsRetryable(err) || ctx.Err() != nil {
			return err
		}
		w.logger.WarnContext(ctx, "spinneret config server unavailable, using local snapshots",
			slog.String("error", err.Error()))
		w.apply(ctx, w.loadSnapshots(), false, true)
		return nil
	}
	if missing := resp.GetMissing(); len(missing) > 0 {
		names := make([]string, len(missing))
		for i, m := range missing {
			names[i] = m.GetGroup() + "/" + m.GetKey()
		}
		w.logger.InfoContext(ctx, "spinneret config items not found", slog.String("items", strings.Join(names, ", ")))
	}
	w.apply(ctx, resp.GetItems(), true, false)
	return nil
}

func (w *ConfigWatcher) loadSnapshots() []*ConfigItem {
	if w.store == nil {
		return nil
	}
	var items []*ConfigItem
	for _, k := range w.keys {
		item, err := w.store.load(k.Group, k.Key)
		if err != nil {
			w.logger.Warn("spinneret ignored config snapshot",
				slog.String("item", k.String()), slog.String("error", err.Error()))
			continue
		}
		if item != nil {
			items = append(items, item)
		}
	}
	return items
}

func (w *ConfigWatcher) watchItems() []*WatchItem {
	w.mu.Lock()
	defer w.mu.Unlock()
	out := make([]*WatchItem, len(w.keys))
	for i, k := range w.keys {
		out[i] = &WatchItem{Group: k.Group, Key: k.Key, Version: max(w.items[k].GetVersion(), 0)}
	}
	return out
}

func (w *ConfigWatcher) run(ctx context.Context) {
	defer func() {
		w.mu.Lock()
		w.stopped = true
		w.broadcastLocked()
		w.mu.Unlock()
		w.client.forgetWatcher(w)
		close(w.done)
	}()
	bo := backoff{initial: w.opts.InitialBackoff, max: w.opts.MaxBackoff}
	failures := 0
	timeoutMs := int32(w.opts.Timeout.Milliseconds())
	for ctx.Err() == nil {
		started := time.Now()
		resp, err := w.client.watchOnce(ctx, &WatchConfigRequest{
			Namespace: w.opts.Namespace,
			Items:     w.watchItems(),
			TimeoutMs: timeoutMs,
		})
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			if ReasonOf(err) == ReasonClientClosed {
				w.logger.Info("spinneret config watcher stopped: the client has been closed")
				return
			}
			delay := bo.delay(failures, w.rnd)
			failures++
			level := slog.LevelWarn
			if !IsRetryable(err) {
				level = slog.LevelError
			}
			w.logger.Log(ctx, level, "spinneret config watch failed, retrying",
				slog.Duration("delay", delay), slog.String("error", err.Error()))
			if !sleepContext(ctx, delay) {
				return
			}
			continue
		}
		failures = 0
		w.apply(ctx, resp.GetItems(), true, false)
		if len(resp.GetItems()) == 0 && time.Since(started) < watchMinPollInterval {
			// A poll without changes must wait for the timeout; guard against
			// a misbehaving proxy answering immediately.
			if !sleepContext(ctx, watchMinPollInterval) {
				return
			}
		}
	}
}

// apply stores the items that differ from the known ones, persists them and
// notifies the listeners.
func (w *ConfigWatcher) apply(ctx context.Context, items []*ConfigItem, persist, fromSnapshot bool) {
	w.mu.Lock()
	var changed []*ConfigItem
	for _, item := range items {
		k := ConfigKey{Group: item.GetGroup(), Key: item.GetKey()}
		if _, ok := w.watched[k]; !ok {
			continue
		}
		if cur, ok := w.items[k]; ok && cur.GetVersion() == item.GetVersion() && cur.GetContent() == item.GetContent() {
			continue
		}
		w.items[k] = item
		w.seq++
		w.changedSeq[k] = w.seq
		changed = append(changed, item)
	}
	w.fromSnapshot = fromSnapshot
	listeners := slices.Clone(w.listeners)
	w.broadcastLocked()
	w.mu.Unlock()
	if persist && w.store != nil {
		for _, item := range changed {
			if _, err := w.store.save(item); err != nil {
				w.logger.WarnContext(ctx, "spinneret could not write config snapshot",
					slog.String("group", item.GetGroup()), slog.String("key", item.GetKey()),
					slog.String("error", err.Error()))
			}
		}
	}
	if ctx.Err() != nil {
		return
	}
	w.dispatching.Store(true)
	defer w.dispatching.Store(false)
	for _, item := range changed {
		for _, fn := range listeners {
			w.notify(ctx, fn, item)
		}
	}
}

func (w *ConfigWatcher) notify(ctx context.Context, fn func(*ConfigItem), item *ConfigItem) {
	defer func() {
		if p := recover(); p != nil {
			w.logger.ErrorContext(ctx, "spinneret config change listener panicked",
				slog.String("group", item.GetGroup()), slog.String("key", item.GetKey()),
				slog.Any("panic", p), slog.String("stack", string(debug.Stack())))
		}
	}()
	fn(item)
}
